control/setup_control.go

```diff
@@ -0,0 +1,418 @@
+	// control config to allow for position control of a component
+	PositionControlUsingTrapz bool
+
+	// SensorFeedback2DVelocityControl adds linear and angular blocks to a control
+	// config in order to use the sensorcontrolled base component for velocity control
+	SensorFeedback2DVelocityControl bool
+
+	// DerivativeType is the type of derivative to be used for the derivative block of a control config
+	DerivativeType string
+
+	// UseCustomeConfig is if the necessary config is not created by this setup file
+	UseCustomConfig bool
+
+	// CompleteCustomConfig is the custom control config to be used instead of the config
+	// created by this setup file
+	CompleteCustomConfig Config
+
+	// NeedsAutoTuning is true when all PID values of a PID block are 0 and
+	// the control loop needs to be auto-tuned
+	NeedsAutoTuning bool
+
+	// LoopFrequency is the frequency at which the control loop should run
+	LoopFrequency float64
+
+	// ControllableType is the type of component the control loop will be set up for,
+	// currently a base or motor
+	ControllableType string
+
+	// NeedsSingleAutoTuning is true when, in the case of multiple PID blocks, only one of
+	// them needs to be auto-tuned
+	NeedsSingleAutoTuning bool
+}
+
+// SetupPIDControlConfig creates a control config.
+func SetupPIDControlConfig(
+	pidVals []PIDConfig,
+	componentName string,
+	options Options,
+	c Controllable,
+	logger logging.Logger,
+) (*PIDLoop, error) {
+	pidLoop := &PIDLoop{
+		Controllable: c,
+		logger:       logger,
+		Options:      options,
+		ControlConf:  Config{},
+		ControlLoop:  nil,
+	}
+
+	// set controlConf as either an optional custom config, or as the default control config
+	if options.UseCustomConfig {
+		pidLoop.ControlConf = options.CompleteCustomConfig
+		for i, b := range options.CompleteCustomConfig.Blocks {
+			if b.Type == blockSum {
+				sumIndex = i
+			}
+		}
+	} else {
+		pidLoop.createControlLoopConfig(pidVals, componentName)
+	}
+
+	// auto tune the control loop if needed
+	if options.NeedsAutoTuning {
+		cancelCtx, cancelFunc := context.WithCancel(context.Background())
+		if err := pidLoop.TunePIDLoop(cancelCtx, cancelFunc); err != nil {
+			return nil, err
+		}
+	}
+
+	if options.NeedsSingleAutoTuning {
+		if err := pidLoop.StartControlLoop(); err != nil {
+			return nil, err
+		}
+	}
+
+	return pidLoop, nil
+}
+
+// TunePIDLoop runs the auto-tuning process for a PID control loop.
+func (p *PIDLoop) TunePIDLoop(ctx context.Context, cancelFunc context.CancelFunc) error {
+	var errs error
+	p.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(func() {
+		defer func() {
+			cancelFunc()
+			if p.ControlLoop != nil {
+				p.ControlLoop.Stop()
+				p.ControlLoop = nil
+			}
+		}()
+		defer p.activeBackgroundWorkers.Done()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		if p.Options.UseCustomConfig {
+			if err := p.StartControlLoop(); err != nil {
+				p.logger.Error(err)
+			}
+			return
+		}
```

Comment:
@JohnN193 does this work you think?

Reply:
ah yes great point

Reply:
done! I think everything should work, but with future PRs these functions will actually be used so I'll do testing and we will know for sure then